src/btree/mod.rs

```diff
@@ -0,0 +1,556 @@
+	}
+
+	pub fn col(&self) -> ColId {
+		self.0 as ColId
+	}
+}
+
+impl std::fmt::Display for BTreeTableId {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		write!(f, "{:02}", self.col())
+	}
+}
+
+#[derive(Clone, PartialEq)]
+pub struct BTreeIndex {
+	pub root: u64,
+	pub depth: u32,
+}
+
+const ORDER: usize = 8;
+const ORDER_CHILD: usize = ORDER + 1;
+
+struct Entry {
+	encoded: LogEntry<Vec<u8>>,
+}
+
+impl Entry {
+	fn empty() -> Self {
+		Self::from_encoded(Vec::with_capacity(ENTRY_CAPACITY))
+	}
+
+	fn from_encoded(enc: Vec<u8>) -> Self {
+		Entry {
+			encoded: LogEntry::new(enc),
+		}
+	}
+
+	fn read_separator(&mut self) -> Option<SeparatorInner> {
+		if self.encoded.offset() == self.encoded.inner_mut().len() {
+			return None;
+		}
+		let value = self.encoded.read_u64();
+		let head = self.encoded.read_slice(1);
+		let head = head[0];
+		let size = if head == u8::MAX {
+			self.encoded.read_u32() as usize
+		} else {
+			head as usize
+		};
+		let key = self.encoded.read_slice(size).to_vec();
+		if value == 0 {
+			return None;
+		}
+		Some(SeparatorInner {
+			key,
+			value,
+		})
+	}
+
+	fn write_separator(&mut self, key: &Vec<u8>, value: u64) {
+		let size = key.len();
+		let inner_size = self.encoded.inner_mut().len();
+		if size >= u8::MAX as usize {
+			self.encoded.inner_mut().resize(inner_size + 8 + 1 + 4 + size, 0); 
+		} else {
+			self.encoded.inner_mut().resize(inner_size + 8 + 1 + size, 0); 
+		};
+		self.encoded.write_u64(value);
+		if size >= u8::MAX as usize {
+			self.encoded.write_slice(&[u8::MAX]);
+			self.encoded.write_u32(size as u32);
+		} else {
+			self.encoded.write_slice(&[size as u8]);
+		}
+		self.encoded.write_slice(key.as_slice());
+	}
+
+	fn read_child_index(&mut self) -> Option<u64> {
+		let index = self.encoded.read_u64();
+		if index == 0 {
+			None
+		} else {
+			Some(index)
+		}
+	}
+
+	fn write_child_index(&mut self, index: u64) {
+		let inner_size = self.encoded.inner_mut().len();
+		self.encoded.inner_mut().resize(inner_size + 8, 0);
+		self.encoded.write_u64(index);
+	}
+
+	fn write_header(&mut self, btree_index: &BTreeIndex) {
+		self.encoded.set_offset(0);
+		self.encoded.inner_mut().resize(8 + 4, 0);
+		self.encoded.write_u64(btree_index.root);
+		self.encoded.write_u32(btree_index.depth);
+	}
+}
+
+pub struct BTreeTable {
+	id: BTreeTableId,
+	tables: RwLock<Vec<ValueTable>>,
+	preimage: bool,
+	ref_counted: bool,
+	compression: Compress,
+}
+
+impl BTreeTable {
+	pub fn open(
+		id: BTreeTableId,
+		values: Vec<ValueTable>,
+		metadata: &crate::options::Metadata,
+	) -> Result<Self> {
+		if let Some(address) = Self::btree_index_address(&values) {
+			let size_tier = address.size_tier() as usize;
+			if !values[size_tier].is_init() {
+				let btree_index = BTreeIndex {
+					root: HEADER_POSITION,
+					depth: 0,
+				};
+				let mut entry = Entry::empty();
+				entry.write_header(&btree_index);
+				values[size_tier].init_with_entry(&*entry.encoded.inner_mut())?;
```

Comment:
Is it really necessary?

Reply:
This initialization is not really good looking, but seems necessary. Value table only really work on existing entries (everything is always indexed from the index table), so I found it easier to just init the header entry when opening the table. The alternative would have been to change the value table code to be safer (check offset limit on entry access), but doing so just for the uninitialized btree header case did not sound good to me.